crates/hir/src/lib.rs

```diff
@@ -2181,6 +2181,53 @@ impl Function {
         db.function_data(self.id).has_async_kw()
     }
 
+    /// Whether this function is a `fn` that returns `impl Future`.
+    pub fn is_desugar_async(self, db: &dyn HirDatabase) -> bool {
+        if self.is_async(db) || self.is_const(db) {
+            return false;
+        }
+
+        let Some(impl_traits) = self.ret_type(db).as_impl_traits(db) else { return false };
+
+        let Some(future_trait_id) =
+            db.lang_item(self.ty(db).env.krate, LangItem::Future).and_then(|t| t.as_trait())
+        else {
+            return false;
+        };
+
+        let Some(size_trait_id) =
+            db.lang_item(self.ty(db).env.krate, LangItem::Sized).and_then(|t| t.as_trait())
+        else {
+            return false;
+        };
+
+        let Some(sync_trait_id) =
+            db.lang_item(self.ty(db).env.krate, LangItem::Sync).and_then(|t| t.as_trait())
+        else {
+            return false;
+        };
+
+        // TODO: There's no `LangItem::Send`. How do we get the id of `Send` trait?
```

Comment:
Help wanted here🙋

Reply:
Huh that is very surprising. Goes to show that the Send/Sync stuff really is more a library level concern than a language one, interesting.

I think we can skip looking for those two all together and instead check whether the impl traits (aside from the principal `Future` one) are auto traits or not. That is more general, as the relevant part about those traits being leaked is them being auto traits, not them being `Send` or `Sync`